superbench/analyzer/diagnosis_rule_op.py

```diff
@@ -0,0 +1,121 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+"""A module for data diagnosis rule ops."""
+
+from typing import Dict, Callable
+
+import pandas as pd
+
+from superbench.benchmarks.context import Enum
+from superbench.common.utils import logger
+
+
+class DiagnosisRuleType(Enum):
+    """The Enum class representing different rule ops."""
+
+    VARIANCE = 'variance'
+    VALUE = 'value'
+
+
+class RuleOp:
+    """RuleOp class to maintain all rule functions."""
+
+    functions: Dict[DiagnosisRuleType, Callable] = dict()
+
+    @classmethod
+    def add_rule_func(cls, rule_type):
+        """Add rule fuction.
+
+        Args:
+            rule_type (DiagnosisRuleType): The type of rule function.
+
+        Return:
+            decorator (Callable): return the decorator to add the rule function.
+        """
+        def decorator(func):
+            cls.functions[rule_type] = func
+            return func
+
+        return decorator
+
+    @classmethod
+    def get_rule_func(cls, rule_type):
+        """Get rule fuction by rule_type.
+
+        Args:
+            rule_type (DiagnosisRuleType): The type of rule function.
+
+        Return:
+            func (Callable): rule function, None means invalid rule type.
+        """
+        if rule_type in cls.functions:
+            return cls.functions[rule_type]
+
+        return None
+
+    @classmethod
+    def variance(cls, data_row, rule, summary_data_row, details, categories):
+        """Rule op function of variance."""
+        pass_rule = True
+        # parse criteria and check if valid
+        if not isinstance(eval(rule['criteria'])(0), bool):
+            logger.log_and_raise(exception=Exception, msg='invalid criteria format')
+        # every metric should pass the rule
+        for metric in rule['metrics']:
+            violate_metric = False
+            # metric not in raw_data not the value is none, miss test
```

Comment:
`# metric not in raw_data not the value is none, miss test` =>
`# metric not in raw_data or the value is none, miss test`